File: tester/java/Tester.java
package tester.java;

import utility.ClassCaller;

import java.io.IOException;
import java.io.Serializable;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map.Entry;

/** A class to compare an assignment's solution output to the expected output of a test case. */
public class Tester {
  /**
   * Run the specified class's specified method and compare the output to the expected output from
   * the list of input files.
   *
   * <p>Output filenames are derived from the input filenames, where the string "input" is replaced
   * by "output."
   *
   * @param args the class name, followed by a list of input files, followed by options:
   *     <ul>
   *       <li>-maxsize:&lt;maximum input size&gt;
   *       <li>-name:&lt;method that returns solution&gt;
   *     </ul>
   */
  public static void main(String[] args) {

    // If the usage is incorrect, display usage string.
    if (args.length < 2) {
      displayUsageString();
      System.exit(0);
    }

    // Parse the arguments.
    int argPointer = 0;

    // Parse the options.
    HashMap<String, String> options = new HashMap<String, String>();
    while (args[argPointer].startsWith("-")) {
      int indexOfColon = args[argPointer].indexOf(":");
      String parameterName = args[argPointer].substring(1, indexOfColon);
      String argument = args[argPointer].substring(indexOfColon + 1, args[argPointer].length());
      options.put(parameterName, argument);
      argPointer++;
    }

    // Parse the solution class name.
    final String solutionClassName = args[argPointer++];

    // Parse the file names.
    ArrayList<String> filenames = new ArrayList<String>();
    while (argPointer < args.length) {
      if (validateFilename(args[argPointer])) {
        filenames.add(args[argPointer]);
      }
      argPointer++;
    }

    // Display feedback to user.
    System.out.println("\nTesting your code's output with the expected output.\n");

    // Display options.
    for (Entry<String, String> entry : options.entrySet()) {
      System.out.println(entry.getKey() + " => " + entry.getValue());
    }
    System.out.println();

    // Retrieve the main method of the solution class.
    String methodName = options.get("name") != null ? options.get("name") : "main";
    Method method = ClassCaller.getMethod(solutionClassName, methodName);

    // Sort the filenames, so simpler problems are processed first.
    filenames.sort(new InputSizeOrder());

    // Initialize an array of failed tests, to be displayed later.
    ArrayList<String> failedTests = new ArrayList<String>();

    // Iterate over all the filenames in the command-line arguments.
    for (String inputFilename : filenames) {
      String[] filenameStringArray =
          inputFilename.substring(0, inputFilename.indexOf(".")).split("_");

      // Determine if this input file exceeds the maxsize.
      if (options.get("maxsize") != null) {
        if (Integer.parseInt(options.get("maxsize")) < Integer.parseInt(filenameStringArray[3])) {
          System.out.print("\033[36m");
          System.out.println("skipped " + inputFilename + "\t-> exceeded maxsize");
          System.out.print("\033[0m");
          continue;
        }
      }

      // Infer the output filename from the input filename.
      String outputFilename = inputFilename.replaceAll("input", "output");

      // Retrieve the expected result.
      String expectedResult = null;
      try {
        expectedResult = new String(Files.readAllBytes(Paths.get(outputFilename)), "UTF-8");
      } catch (IOException exception) {
        exception.printStackTrace();
      }

      // Get the result of calling the solution class's main() method on this file.
      String[] thisFilenameAsArray = {inputFilename};
      final String result = ClassCaller.callMethod(method, thisFilenameAsArray);

      // Display whether or not this test matches the expected results.
      if (expectedResult != null) {
        if (result.trim().equals(expectedResult.trim())) {
          System.out.print("\033[32m");
          System.out.println(inputFilename + "\t✔");
          System.out.print("\033[0m");
        } else {
          System.out.print("\033[31m");
          System.out.println(inputFilename + "\t✘");
          System.out.println("\tExpected: " + expectedResult.trim());
          System.out.println("\t  Result: " + result.trim());
          System.out.print("\033[0m");
          failedTests.add(inputFilename);
        }
      }
    }

    // Display summary of test results.
    System.out.println();
    if (failedTests.isEmpty()) {
      System.out.print("\033[32m");
      System.out.println("✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔");
      System.out.println("All tests passed");
      System.out.println("✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔");
      System.out.print("\033[0m");
    } else {
      System.out.print("\033[31m");
      System.out.println("✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘");
      System.out.println("Some tests didn't pass");
      System.out.println("✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘✘");
      System.out.println("\nFailed tests:");
      for (String failedTest : failedTests) {
        System.out.println("\t" + failedTest);
      }
      System.out.print("\033[0m");
    }
  }

  /**
   * Return whether or not a filename is valid.
   *
   * @param string the filename to be validated
   * @return whether or not the filename is valid
   */
  private static boolean validateFilename(String string) {
    if (string.split("_").length == 4) {
      return true;
    }
    System.out.println(
        "'"
            + string
            + "' must match the pattern '\033[1minput_description_index_problemSize.txt\033[0m'");
    return false;
  }

  /** Display the class's usage string. */
  private static void displayUsageString() {
    System.out.println(
        "Usage:"
            + "java -cp <path to tester class>:<path to solution class> "
            + Tester.class.getName()
            + "  [-options] <solution class name> <list of input files>");
    System.out.println("where options include:");
    System.out.println("\t-maxsize        maximum input size");
    System.out.println("\t-name           method that returns solution (default: main)");
  }

  /**
   * A comparator to sort filenames based on the problem size.
   *
   * @author beaunus
   */
  private static class InputSizeOrder implements Comparator<String>, Serializable {

    /** Auto-generated serialVersionUID. */
    private static final long serialVersionUID = 1L;

    @Override
    public int compare(String o1, String o2) {
      String[] o1AsArray = o1.split("_");
      String[] o2AsArray = o2.split("_");

      int o1Size = Integer.parseInt(o1AsArray[o1AsArray.length - 2].split("\\.")[0]);
      int o2Size = Integer.parseInt(o2AsArray[o2AsArray.length - 2].split("\\.")[0]);

      return o1Size - o2Size;
    }
  }
}

File: tester/python3/tester.py
import sys
import os
import importlib

def test(student_algorithm, test_cases_folder, name='alg', max_size=-1, only=[]):
	# import the algorithm
	*alg_path, alg_file = student_algorithm.split('/')
	sys.path.append('/'.join(alg_path))
	alg = importlib.import_module(alg_file[:-3])
	alg = getattr(alg, name)

	# print the options to the user
	print('Testing algorithm at "'+ student_algorithm + '".')
	if name != 'alg' or max_size != -1 or only != []:
		print('Options: ')
		if name != 'alg': print('\tname =', name)
		if max_size != -1: print('\tmax_size =', max_size)
		if len(only) > 0:
			print('\tonly =')
			for only_file in only:
				print('\t\t' + only_file)

	# read the test cases titles
	test_cases = []
	for filename in os.listdir(test_cases_folder):
		if filename[:5] != 'input':
			continue

		# get all input files
		file_info = filename[:-4].split('_')
		# First three parts are type, user, and number
		file_type, file_user, file_num = file_info[:3]
		# Rest parts are input sizes
		file_size = 1
		for size in file_info[3:]:
			file_size *= int(size)

		fileref = file_info[1:]
		fileref_str = '_'.join([str(x) for x in fileref])

		# filter 'only' files
		if len(only)>0 and fileref_str not in only:
			print('skipped', fileref_str, '\t-> not in only')
			continue

		# if the size is bigger than the max_size option, don't run that test
		if max_size != -1 and file_size > max_size:
			print('skipped', fileref_str, '\t-> exceeded max_size')
			continue
		
		# construct the output
		output_file = (test_cases_folder + '/output_' + fileref_str + '.txt')
		# read the expected output
		with open(output_file, 'r') as ofile:
			output = []
			for line in ofile:
				output.append(line.strip())

		# construct an array with the relation between file and test case solution
		test_cases.append([file_size, fileref, output])

	# sort by size to run first the samller tests
	test_cases.sort(key=lambda x: x[0])

	# run the tests
	print()
	tests_passed = 0
	tests_total = 0
	failed_tests = []
	for file_size, fileref, expected_output in test_cases:
		fileref_str = '_'.join(str(x) for x in fileref)
		filename = test_cases_folder + '/input_' + fileref_str + '.txt'

		alg_output_unprocessed = alg(filename)
		if not isinstance(alg_output_unprocessed, list):
			alg_output_unprocessed = [alg_output_unprocessed]

		alg_output = [str(x) for x in alg_output_unprocessed]

		if len(alg_output) != len(expected_output):
			print('Error. Your algorithm should be returning', len(expected_output),
				'values, but it is returning', len(alg_output), 'instead. Fix this ' +
				'and run the tester again.')
			return

		answers_ok = []
		print(fileref_str, end='\t')
		errors_buffer = ''
		for i in range(len(alg_output)):
			if expected_output[i] == alg_output[i]:
				answers_ok.append(True)
				print(u'\N{heavy check mark}', end='')
			else:
				answers_ok.append(False)
				print(u'\N{heavy ballot x}', end='')
				errors_buffer += ('\n    Answer ' + str(i+1) + ':\n\tExpected: ' + expected_output[i] +
					'\n\tResult: ' + alg_output[i])
				failed_tests.append(fileref_str)

		if all(answers_ok):
			tests_passed += 1
			print()
		else:
			print(errors_buffer)
			errors_buffer = ''

		tests_total += 1

	print()
	if tests_total == tests_passed:
		message = 'All tests passed'
		print(u'\N{heavy check mark}'*len(message))
		print(message)
		print(u'\N{heavy check mark}'*len(message))
	else:
		message = 'Some tests didn\'t pass'
		print(u'\N{heavy ballot x}'*len(message))
		print(message)
		print(u'\N{heavy ballot x}'*len(message))
		print('Failed tests:', ','.join(failed_tests))


if __name__ == '__main__':
	if (len(sys.argv) < 3):
		print('usage: ' + __file__ + 
				' path/to/your/algorithm.py path/to/testcases [options]\n')
		print('valid options: ')
		print('\tmax_size=[maximum input size]')
		print('\tname=[function that returns solution]')
		print('\tonly=[specific tests to run, excluding others]')
		sys.exit()
	tester_name, student_algorithm, test_cases_folder, *options = sys.argv
	
	valid_options = ['max_size', 'name', 'only']
	name = 'alg'
	max_size = -1
	only = []
	for opt in options:
		opt_name, opt_value = opt.split('=')
		if opt_name == 'max_size':
			max_size = int(opt_value)

		if opt_name == 'name':
			name = opt_value

		if opt_name == 'only':
			only = opt_value.split(',')

		if opt_name not in valid_options:
			print('Unknown option "' + opt_name + 
					'". Valid options are\n\t"name"\n\t"max_size"')
			print('Option "'+ opt_name +'" not set')

	option_values = [name, max_size, only]
	test(student_algorithm, test_cases_folder,  *option_values)
